api/src/main/java/org/apache/iceberg/expressions/BoundLiteralPredicate.java

```diff
@@ -69,6 +69,8 @@ public boolean test(T value) {
         return cmp.compare(value, literal.value()) != 0;
       case STARTS_WITH:
         return String.valueOf(value).startsWith((String) literal.value());
+      case NOT_STARTS_WITH:
+        return !String.valueOf(value).startsWith((String) literal.value());
```

Comment:
Should we be using the predefined comparators here? This seems like it could get iffy with non-ascii values given the difference between UTF-8 and Java's default String encoding of UTF-16.

I chose to go with the existing style here, but I'm not 100% convinced this is the best solution as opposed to using a comparator.

Reply:
The other way to handle this to be similar to the rest of the cases and to use the `Comparator` that `literal` brings with it would be to convert the `T` to a ByteBuffer, and then truncate it and check equality with the ByteBuffer representation of `literal`, as follows.

```java
ByteBuffer valueAsBytes = Conversions.toByteBuffer(ref().type(), value);
ByteBuffer literalAsBytes = literal.toByteBuffer();
int length = Math.min(valueAsBytes.remaining(), literalAsBytes.remaining());
Comparators.unsignedBytes(BinaryUtil.truncateBinary(valueAsBytes, length), prefixAsBytes)) == 0;
```

Or if we wanted to just use the comparator from literal, we might consider the following. However, for `CharSequence` etc we would need to truncate like we do elsewhere so I don't think that this would be correct.

```java
cmp.compare(value, literal.value()) == 0;
```

Reply:
It's also tempting to just use the comparator from literal, like follows:

```java
cmp.compare(value, literal.value()) == 0;
```

However, because we're only interested in checking whether or not `value` does not start with the predicate's literal term, we would need to truncate like we do elsewhere so I don't think that the simple one line above would be correct (elegant as it would be).

Reply:
I think alignment with the previous code is find. I would be worried about any changes to past matching behavior and I think one of the concerns here is that there are multiple groups of bytes which may represent the same string so you can't just compare the binary values I think....

Reply:
Cool. I agree about being worried re: changing past behavior. Wasn't entirely sure of all of the conditions around it (I'm not a UTF-8 expert admittedly), but wanted to bring it up. I'll close this comment then! Thanks for clarification.